_docs/cardano/2017-01-06-updating.md

```diff
@@ -157,93 +154,96 @@ modifications to the protocol with some degree of concurrency.
+parts: a stakeholder with a stake large enough could update and maintain their
+own chain, rejecting blocks from others, but others would be still able to
+maintain their chain, rejecting blocks from this stakeholder (since he does not
+have a majority of stake, and therefore cannot catch up with the rest of the
+crowd, so his chain is shorter). A simple resolution rule could go like this: if
+95% of the latest 2016 blocks have a newer block version, the blocks with the
+older version are rejected.
+
+It may seem unclear why we would like to make some block version invalid at some
+moment. The key insight here is that a new feature is actually a restriction on
+what we had had previously. For example, currently we have plain old
+transactions which may contain either PublicKey-based addresses or Script-based
+ones. Then at some point we decide to include a third address type (no matter
+what its purpose is). Which strategy do we need for verifying a block with a
+transaction with an address of unknown type? Obviously the only option is not
+verifying this address.
+
+And then imagine somebody proposing a transaction to this address, possibly
+doing that with an intention to secure funds from being spent until some
+conditions are met — and then watching them being spent in some other
+transaction in a block with version 1. This is the point. We cannot make use of
+a restriction without waiting for the network to start assuming the old version
+to be deprecated (since we can only start rejecting blocks when their version is
+deprecated).
+
+Another problem is that in our implementations all blocks have a field called
+[`attributes`](https://github.com/input-output-hk/cardano-sl/blob/db0746e4693b3be645959d42d808ced3ee03f402/src/Pos/Types/Block/Types.hs#L87)
+which is used for storing auxiliary information. An attacker could claim she
+uses a higher version of the protocol and generate a block whose `attributes`
+would be polluted with meaningless keys. We would have to accept it, and that
+would bloat the blockchain.
+
+This is the motivation for the logic described below.
+
+In our implementation, block version can be in the following states:
+
+* **Adopted**, when the softfork resolution rule (see below) for a confirmed
+  block version is triggered.
+* **Confirmed**, when there is an update proposal that contains a confirmed
+  version of the software and this block version. Note that "confirmed version of
+  the software" is a technical term defined elsewhere. If there are multiple block
+  versions such that corresponding software is confirmed, but these versions
+  aren't adopted, we call them **competing**. For instance, there might be
+  versions `2.0.0`, `2.0.1`, `1.2.0`, `1.2.1`, `1.1.1` and `1.1.2`, with the last
+  adopted version being `1.1.3`. In this case the versions that are competing are
+  `2.0.0`, `2.0.1`, `1.2.0` and `1.2.1`. `1.1.1` and `1.1.2` aren't competing,
```

Comment:
It's unclear where your enumeration of competing versions starts and where it ends.